### DISCUSSION TITLE
API for returning images and using Next/image, stalled requests

### DISCUSSION THREAD

**Discussion by szanuje** [vercel/next.js#39932] [Help] [answered]
Hey, I am trying to set up an API for fetching images from third party source and serving them as a NextJS API.

I am using Next/image to show images in React client application. E.g.

```
          <Image
            src={`/api/avatar/${name}`}
            layout="fill"
            alt={name}
            onLoadingComplete={handleLoading}
            onError={handleError}
          />
```
Note that I need to handle the case when image is not present.

My serverless function looks like that:

```
const handler = async (req: NextApiRequest, res: NextApiResponse) => {
  const { name } = req.query;

  try {
    res.setHeader(
      "Cache-Control",
      `public, immutable, no-transform, s-maxage=31536000, max-age=31536000`,
    );
    res.setHeader("content-type", "image/png");

    const { url } = (await getAvatar(name as string)) || {};

    if (!url) return res.status(204).end();

    const avatarResponse = await fetch(url);

    if (!avatarResponse.ok) return res.status(204).end();

    const contentType = avatarResponse.headers.get("content-type");
    const body = avatarResponse.body;

    res.setHeader("content-type", contentType);

    res.status(200).send(body);
  } catch (error) {
    console.error({ name, avatarApiError: error.message });
    res.status(204).end();
  }
};

export default withSentry(handler);
```

I tried to use multiple variations of status codes and responses, but I very often get:
`API resolved without sending a response for /api/avatar/wzz.eth, this may result in stalled requests.`
In some scenario I also observed that Network tab shows `304` response for that API when loading image.

Next version: `12.2.5`
Sentry: `"@sentry/nextjs": "7.11.1"`

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I take it that, removing sentry does remove the issue? 

I'd try adding this to the file, https://nextjs.org/docs/api-routes/request-helpers#custom-config, while keeping `withSentry`:

```js
export const config = {
  api: {
    externalResolver: true,
  },
}
```

Does that help out?

**Reply by szanuje** [reply]
@icyJoseph Thank you for quick response. Unfortunately removing Sentry wrapper `withSentry()` or/and exporting `config` didn't work.

**Reply by icyJoseph** [reply]
Ah, I meant to keep the wrapper! Sorry about that! it is necessary to have the `withSentry` function, because, that's the external resolver...

**Reply by icyJoseph** [reply]
I'd also add `return`, like `return res.status(200).send(body)`, and so on.

**Reply by icyJoseph** [reply]
And also, there's a limit to the payload transferred through API routes. I reckon perhaps you are hitting that limit, and the handler just crashes out, and nobody is there to resolve the crash? https://nextjs.org/docs/messages/api-routes-response-size-limit

**Reply by szanuje** [reply]
@icyJoseph This is the current code:
```
const handler = async (req: NextApiRequest, res: NextApiResponse) => {
  const { name } = req.query;

  try {
    res.setHeader(
      "Cache-Control",
      `public, immutable, no-transform, s-maxage=31536000, max-age=31536000`,
    );
    res.setHeader("content-type", "image/png");

    const { url } = (await getAvatar(name as string)) || {};

    if (!url) return res.status(404).end();

    const avatarResponse = await fetch(url);

    if (!avatarResponse.ok) return res.status(404).end();

    const contentType = avatarResponse.headers.get("content-type");
    const body = avatarResponse.body;

    res.setHeader("content-type", contentType);

    return res.status(200).send(body);
  } catch (error) {
    console.error({ name, avatarApiError: error.message });
    return res.status(404).end();
  }
};

export default withSentry(handler);

export const config = {
  externalResolver: true,
};
```
I changed the response when there's not image to `404`, because when it was `204` I could't enter the page manually (browser request was cancelled immediately).
I added `return` statements everywhere and kept the Sentry wrapper.
Payloads are most often 0 bytes currently because there are very few images being available. In different case they're 300B.
I still get the stalled request warning, for every component that is being loaded which uses `<Image .. /> component.

**Reply by icyJoseph** [reply]
Do you think you could fix a [stackblitz](https://stackblitz.com/) reproduction demo? 

So In short every component that uses the Image component, ends up with the stalled request warning? Even when it succeds?

**Reply by szanuje** [reply]
@icyJoseph When image succeeds, there's no warning.

stakblitz: https://stackblitz.com/edit/nextjs-n3g8tk?file=pages/index.js

Press a button to load image from API endpoint. Each time new query path is generated.

**Reply by icyJoseph** [reply]
Hi,

I think I made a mistake with the initial snippet...

```ts
export const config = {
  api: {
    externalResolver: true,
  },
}
```

Notice the nested `api` key inside config... sorry about that, I've also updated my main comment.

**Reply by szanuje** [reply]
@icyJoseph Seems like it fixed the issue. Thank you so much!

**Comment by SH20RAJ** [comment]
For Latest App Router

```
import { NextResponse } from 'next/server';
import prisma from '@/lib/prisma'; // Adjust the import according to your project structure

async function getImageData(imageUrl) {
  try {
    const response = await fetch(imageUrl);
    if (!response.ok) {
      throw new Error(`Failed to fetch image: ${response.statusText}`);
    }
    const buffer = await response.arrayBuffer();
    return Buffer.from(buffer);
  } catch (error) {
    console.error('Error fetching image data:', error);
    throw error;
  }
}

export async function GET(req) {
  const { id } = req.params;

  try {
    const image = await prisma.image.findUnique({
      where: {
        id: parseInt(id, 10),
      },
    });

    if (!image) {
      return NextResponse.json({ error: 'Image not found' }, { status: 404 });
    }

    const imageData = await getImageData(image.url);
    const mimeType = image.mimeType;

    return new NextResponse(imageData, {
      headers: {
        'Content-Type': mimeType,
      },
    });
  } catch (error) {
    console.error('Error in GET handler:', error);
    return NextResponse.json({ error: 'Internal Server Error' }, { status: 500 });
  }
}

```
